Applications/Spire/Source/Ui/DropDownList.cpp

```diff
@@ -0,0 +1,255 @@
+    const HighlightedSignal::slot_type& slot) const {
+  return m_highlighted_signal.connect(slot);
+}
+
+connection DropDownList::connect_index_selected_signal(
+    const IndexSelectedSignal::slot_type& slot) const {
+  return m_index_selected_signal.connect(slot);
+}
+
+connection DropDownList::connect_value_selected_signal(
+    const ValueSelectedSignal::slot_type& slot) const {
+  return m_value_selected_signal.connect(slot);
+}
+
+QVariant DropDownList::get_value(int index) {
+  if(m_layout->count() - 1 < index) {
+    return QVariant();
+  }
+  return get_widget(index)->get_value();
+}
+
+void DropDownList::insert_item(DropDownItem* item) {
+  m_item_selected_connections.push_back(item->connect_selected_signal(
+    [=] (auto value) {
+      on_item_selected(std::move(value), item);
+    }));
+  m_layout->insertWidget(m_layout->count(), item);
+  update_height();
+}
+
+int DropDownList::item_count() const {
+  return m_layout->count();
+}
+
+void DropDownList::remove_item(int index) {
+  if(index > m_layout->count() - 1) {
+    return;
+  }
+  m_item_selected_connections.erase(
+    m_item_selected_connections.begin() + index);
+  auto layout_item = m_layout->takeAt(index);
+  delete layout_item->widget();
+  delete layout_item;
+  update_height();
+}
+
+bool DropDownList::set_highlight(const QString& text) {
+  for(auto i = 0; i < item_count(); ++i) {
+    if(m_item_delegate.displayText(get_value(i)).startsWith(text,
+        Qt::CaseInsensitive)) {
+      set_highlight(i);
+      return true;
+    }
+  }
+  return false;
+}
+
+void DropDownList::set_items(std::vector<DropDownItem*> items) {
+  while(auto item = m_layout->takeAt(0)) {
+    delete item->widget();
+    delete item;
+  }
+  m_item_selected_connections.clear();
+  for(auto i = std::size_t(0); i < items.size(); ++i) {
+    m_layout->addWidget(items[i]);
+    m_item_selected_connections.push_back(items[i]->connect_selected_signal(
+      [=, item = items[i]] (auto value) {
+        on_item_selected(std::move(value), item);
+      }));
+  }
+  if(m_layout->count() > 0) {
+    update_height();
+    static_cast<DropDownItem*>(m_layout->itemAt(0)->widget())->set_highlight();
+    m_highlight_index = 0;
+  } else {
+    hide();
+  }
+}
+
+DropDownItem* DropDownList::get_widget(int index) {
+  return static_cast<DropDownItem*>(m_layout->itemAt(index)->widget());
+}
+
+void DropDownList::activate_next() {
+  if(m_layout->count() == 0) {
+    return;
+  }
+  if(!m_highlight_index) {
+    set_highlight(0);
+    return;
+  }
+  set_highlight((*m_highlight_index) + 1);
+}
+
+void DropDownList::activate_previous() {
+  if(m_layout->count() == 0) {
+    return;
+  }
+  if(!m_highlight_index) {
+    set_highlight(m_layout->count() - 1);
+    return;
+  }
+  set_highlight((*m_highlight_index) - 1);
+}
+
+void DropDownList::set_highlight(int index) {
+  if(m_highlight_index) {
+    get_widget(*m_highlight_index)->reset_highlight();
+  }
+  if(index < 0) {
+    index = m_layout->count() - 1;
```

Comment:
Not a fan of this approach. You are trying to accommodate ```activate_previous``` and ```activate_next``` with this implementation. Instead, ```set_highlight``` should strictly apply a highlight on the item at the specified ```index```. It is up to ```activate_XXX``` to perform the "wrapping" logic or any other logic they need to determine what the "next" item is. It's not the job of ```set_highlight``` to perform this.